src/Covenant/Test.hs

```diff
@@ -69,27 +126,435 @@ instance Arbitrary Concrete where
+
+   When run multiple times in the monadic context, will reuse single variable declarations that are "in scope" (i.e. have already been generated and are
+   known in the DataGenM state).
+
+   TODO: Rework this to generate declarations with an arbitrary number of tyvar arguments. Doing so would be fairly simple (but isn't needed ATM)
+-}
+genPolymorphic1Decl :: DataGenM Polymorphic1
+genPolymorphic1Decl =
+  Polymorphic1 <$> do
+    tyNm <- freshTyName
+    logArity tyNm count1
+    numCtors <- GT.liftGen $ chooseInt (1, 5)
+    polyCtors <- concat <$> GT.vectorOf numCtors (genPolyCtor tyNm)
+    let result = Covenant.Internal.Type.DataDeclaration tyNm count1 (Vector.fromList polyCtors)
+    returnDecl result
+  where
+    -- We return a single constructor UNLESS we're generating a recursive type, in which case we have to return 2 to ensure a base case
+    genPolyCtor :: Covenant.Internal.Type.TyName -> DataGenM [Covenant.Internal.Type.Constructor AbstractTy]
+    genPolyCtor thisTy = do
+      ctorNm <- freshConstructorName
+      numArgs <- GT.liftGen $ chooseInt (1, 5)
+      argsRaw <- GT.vectorOf numArgs polyArg
+      let recCase = Covenant.Internal.Type.Datatype thisTy (Vector.singleton (Covenant.Internal.Type.Abstraction (BoundAt Z ix0)))
+      if recCase `elem` argsRaw
+        then do
+          baseCtorNm <- freshConstructorName
+          let baseCtor = Covenant.Internal.Type.Constructor baseCtorNm mempty
+              recCtor = Covenant.Internal.Type.Constructor ctorNm (Vector.fromList argsRaw)
+          pure [baseCtor, recCtor]
+        else pure [Covenant.Internal.Type.Constructor ctorNm (Vector.fromList argsRaw)]
+      where
+        arityOne :: Count "tyvar" -> Bool
+        arityOne c = c == count1
+
+        polyArg :: DataGenM (Covenant.Internal.Type.ValT AbstractTy)
+        polyArg = do
+          -- first we choose a type with an arity >=1. We have to have at least one of those because we've added the parent type to the arity map
+          availableArity1 <- gets (fmap fst . (filter (arityOne . snd) . M.toList) . view dgArities)
+          someTyCon1 <- GT.elements availableArity1
+          GT.oneof
+            [ pure $ Covenant.Internal.Type.Abstraction (BoundAt Z ix0),
+              pure $ Covenant.Internal.Type.Datatype someTyCon1 (Vector.singleton (Covenant.Internal.Type.Abstraction (BoundAt Z ix0))),
+              GT.liftGen (coerce <$> arbitrary @Concrete)
+            ]
+
+{-
+   Misc Helpers and the Arbitrary instances
+-}
+
+{- This saves us from having to write *another* newtype for each set of a "flavor" of datatypes and doesn't have any real drawbacks -}
+data DataDeclFlavor
+  = ConcreteDecl
+  | ConcreteNestedDecl
+  | SimpleRecursive
+  | Poly1
+
+-- can't be a Set b/c don't have the Ord instances, don't see what benefit a Vector gives since we'll almost always consume this
+-- via folding anyway
```

Comment:
Couldn't you just derive `Ord`?

Reply:
Maybe? I don't remember why I didn't do that. Possibly I ran into something where it wanted me to write an instance? 

But at any rate, it doesn't really _matter_ because: 1) Our machinery ensures that they're all unique, 2) Moreover, our machinery ensures that a bunch of relevant structural components (like, all the names) are also unique, which is what we care about. 

And anyway, if we want something other than a list here, it's probably a *Map* and not a *Set*. We'll only ever look things up by type name. IIRC a Set would be strictly worse than a list here because it's (marginally) more costly to fold over. I'll change this later if the need arises. We don't need any instances for the Map.